rfc-111-standardise-on-one-cli-tool.md

```diff
@@ -0,0 +1,148 @@
+# 111: Standardise on one CLI tool
+
+## Summary
+
+There are currently 4 different CLI tools in use for performing day to
+day tasks relating to connecting to machines and accessing AWS. Three
+of these are maintained by GOV.UK and one by RE.
+
+This RFC seeks to determine which one GOV.UK should standardise on.
+
+## Problem
+
+There exist 4 different internally-developed CLI tools within GDS,
+three of which are maintained by members of GOV.UK and in daily use.
+The sections below summarise what each of these tools are capable of
+doing and how easily installable they are.
+
+### `govuk` and associated scripts from [alphagov/govuk-guix](https://github.com/alphagov/govuk-guix)
+
+- Allows people to connect to machines and services in AWS and Carrenza with `govuk connect`.
+- Contains `govuk aws` for connecting to services and accounts in AWS with a specific `--profile` argument.
+- Contains `govuk data` for replicating data. It requires `govuk-guix` to be set up in some capacity.
+- Written in Ruby.
+- Installable by copying or symlinking the scripts from `bin/` to `$PATH`.
+
+### `govuk` and associated scripts from [alphagov/govuk-cli](https://github.com/alphagov/govuk-cli)
+
+- Contains `govuk connect` and `govuk config`, copied from the above `alphagov/govuk-guix` repo.
+- The migration between `govuk-guix` and `govuk-cli` repos is [ongoing](https://github.com/alphagov/govuk-guix/issues/36), so the two might not be in sync.
+- Written in Ruby.
+- Installable with `brew install alphagov/gds/govuk-cli`.
+
+### `govukcli` from [alphagov/govuk-aws](https://github.com/alphagov/govuk-aws/tree/master/tools/govukcli)
+
+- Allows people to assume role into AWS with `govukcli aws`.
+- Allows people to SSH into machines with `govukcli ssh`.
+- Access is available to all GOV.UK environments due to the `govukcli set-context` command - it is not AWS-specific.
+- Written in Bash.
+- Installable by following [these instructions in the Developer Docs](https://docs.publishing.service.gov.uk/manual/howto-ssh-to-machines-in-aws.html#setup).
+
+### `gds-cli` from [alphagov/gds-cli](https://github.com/alphagov/gds-cli)
+
+- Allows users to assume role into AWS with `gds aws govuk-<environment>`. Can also open links to common tools that teams use (ie team manual, Trello, etc).
+- Deals with MFA transparently using `aws-vault` under the hood.
+- Written in Go.
+- Installable with `brew install alphagov/gds/gds-cli`.
+
+## Proposal
+
+Embrace `gds-cli` as a tool and deprecate any other command-line tools
+that duplicate or have significant convergeance with the functionality
+`gds-cli` supports (ie, assuming roles in AWS).
```

Comment:
My current views on developer tooling in this general area is that there are lots of unknowns at the moment.

I don't know if developers on GOV.UK installing the `govuk-cli` through Homebrew is going to work out, and I don't know how many developers have successfully used the gds-cli AWS functionality, or how it works myself as I haven't managed to get it to work yet.

Also, it isn't clear to me what this proposal is proposing to change, or how. I've been thinking about looking at deprecating the `govukcli` script, but in my mind, that can't happen until it's known how many people are currently using it, and there's a clear suggestion of what to use instead. That's something I've been working towards, but I don't feel even ready to start suggesting that it's deprecated yet.

Reply:
> I don't know if developers on GOV.UK installing the govuk-cli through Homebrew is going to work out, and I don't know how many developers have successfully used the gds-cli AWS functionality, or how it works myself as I haven't managed to get it to work yet.

Why wouldn't it work out? On a sample size of "a lot of Verify and most of TechOps", it has worked OK. If it doesn't, for non-standard setup reasons or access reasons or other reasons [people should raise issues](https://github.com/alphagov/gds-cli/issues) (either there or in the [homebrew-gds repo](https://github.com/alphagov/homebrew-gds/issues) and work things out in the open to encourage learning.

> Also, it isn't clear to me what this proposal is proposing to change, or how.

I could put together more words to document my vision of _how_ this will change things, how we'll achieve the suggested proposal (that is, [this](https://github.com/alphagov/govuk-rfcs/blob/111-standardise-on-one-cli-tool/rfc-111-standardise-on-one-cli-tool.md#proposal)). But equally it's up to everyone to weigh in, so...

Reply:
> > I don't know if developers on GOV.UK installing the govuk-cli through Homebrew is going to work out, and I don't know how many developers have successfully used the gds-cli AWS functionality, or how it works myself as I haven't managed to get it to work yet.
> 
> Why wouldn't it work out? On a sample size of "a lot of Verify and most of TechOps", it has worked OK.

I don't know. Speaking personally, I haven't got it to work yet, and I haven't heard of any developers on GOV.UK using the gds-cli yet. Also, I don't know if the way Verify and TechOps interact with it is consistent with the way developers on GOV.UK might.

I think of trying to popularise it's use as a prerequisite to proposing that it's embraced, and especially for deprecating other tooling.

> > Also, it isn't clear to me what this proposal is proposing to change, or how.
> 
> I could put together more words to document my vision of how this will change things, how we'll achieve the suggested proposal (that is, this). But equally it's up to everyone to weigh in, so...

Maybe less words would help more. The title is quite concise, but to explore it, does "standardise on one CLI tool" mean getting rid of anything not within the `gds-cli` (if that is the one and only tool), like the connect functionality in the govuk-cli repository (as it's not part of the `gds-cli`)?

On the subject of the new `govuk-cli` repository, this RFC seems to be proposing to retire it ("Don't attempt to have a single unified govuk cli"), as what I'm attempting to do with that repository is at least and provide one clear place where scripts relating to GOV.UK can be, regardless of the implementation or function.